src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,208 @@
+    linear_interpolation((psp.rgrid[1:ir_cut],), psp.projs[l+1][i])(r) / r
+end
+
+"""
+    eval_psp_projector_fourier(psp::PspUPF, i::Number, l::Number, q::Number)
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l in at k-space distance q.
+
+!!! Note
+UPFs store `rgrid[i] * β[l,n](rgrid[i])`, so the integrand has `r` instead of `r^2`
+"""
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
+    if iszero(q)
+        if iszero(l)
+            s = one(T)
+        else
+            s = zero(T)
+        end
+    else
+        l_val = Val(l)
+        s = psp_projector_fouier_inner_loop(l_val, psp, i, l, q)
+        s = 2T(π) * s  # 2π because Ry -> Ha included
+    end
+    return s
+end
+
+function psp_projector_fouier_inner_loop(order::Val, psp::PspUpf, i, l, q::T)::T where {T <: Real}
+    N = length(psp.rprojsdr[l+1][i])
+    s = zero(T)
+    @inbounds for ir = 1:N
+        s += sphericalbesselj(order, q * psp.rgrid[ir]) * psp.rprojsdr[l+1][i][ir]
+    end
+    return s
+end
+
+"""
+     eval_psp_local_real(psp::PspUpf, r::Number)
+
+ Evaluate the local potential at real-space distance `r` via linear interpolation on the
+ real-space mesh.
+ """
+function eval_psp_local_real(psp::PspUpf, r::T) where {T <: Real}
+    linear_interpolation((psp.rgrid,), psp.vloc)(r) / 2  # Ry -> Ha
+end
+
+"""
+    eval_psp_local_fourier(psp::PspUpf, q::Number)
+
+Evaluate the local potential at fourier-space distance `q` using a Coulomb correction term
+`-Z/r`.
+"""
+function eval_psp_local_fourier(psp::PspUpf, q::T)::T where {T <: Real}
+    Z = 2 * psp.Zion  # for some strange reason
+    N = length(psp.rvlocpzdr)
+    s = zero(T)
+    @inbounds @fastmath for ir = 1:N
+        # sin(qr) / qr * (V(r) - -Z/r) r^2 dr
+        s += sin(q * psp.rgrid[ir]) * psp.rvlocpzdr[ir]
+    end
+    2T(π) * (s - Z/q) / q  # 2π because Ry -> Ha included
+end
+
+"""
+    eval_psp_energy_correction(T::Type, psp::PspUpf, n_electrons::Number)
+
+Evaluate the pseudopotential energy correction, i.e. the limit of the Fourier-transform of the
+local potential as q -> 0.
+"""
+function eval_psp_energy_correction(T, psp::PspUpf, n_electrons)
+    N = length(psp.rvlocpzdr)
+    s = zero(T)
+    @inbounds for ir = 1:N
+        # (V(r) - -Z/r) r^2 dr
+        s += psp.rgrid[ir] * psp.rvlocpzdr[ir]
+    end
+    2T(π) * n_electrons * s  # 2π because Ry -> Ha included
+end
+
+@inline @fastmath function sphericalbesselj(order::Val{0}, x)
+    sin(x) / x
+end
+@inline @fastmath function sphericalbesselj(order::Val{1}, x)
+    (sin(x) / x - cos(x)) / x
+end
+@inline @fastmath function sphericalbesselj(order::Val{2}, x) 
+    (sin(x) * (3/x - x) - 3cos(x)) / x^2
+end
+@inline @fastmath function sphericalbesselj(order::Val{3}, x)
+    (sin(x) * (15/x - 6x) + cos(x) * (x^2 - 15)) / x^3
+end
+@inline @fastmath function sphericalbesselj(order::Val{4}, x)
+    (sin(x) * (105 - 45x^2 + x^4) + cos(x) * (10x^3 - 105x)) / x^5
+end
```

Comment:
One more thing: These should go to their separate file, similar to the spherical harmonics.